src/main/java/core/basesyntax/Car.java

```diff
@@ -1,13 +1,78 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
 /**
  * Make this class immutable. See requirements in the README file
  */
-public class Car {
-    private int year;
-    private String color;
-    private List<Wheel> wheels;
-    private Engine engine;
+public final class Car {
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = clonedWheels(wheels);
+        this.engine = engine != null ? engine.clone() : null;
+    }
+
+    public int getYear() {
+        return year;
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public List<Wheel> getWheels() {
+        return clonedWheels(wheels);
+    }
+
+    public Engine getEngine() {
+        return engine != null ? engine.clone() : null;
+    }
+
+    public Car addWheel(Wheel wheel) {
+        List<Wheel> addedWheels = new ArrayList<>(wheels);
+        addedWheels.add(wheel);
+        return new Car(year, color, addedWheels, engine.clone());
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(year, color, wheels, engine);
+    }
+
+    public Car changeColor(String color) {
+        return new Car(year, color, wheels, engine.clone());
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(year, color, engine);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        Car car = (Car) obj;
+        return year == car.year && Objects.equals(color,car.color)
+                && Objects.equals(engine,car.engine);
+    }
+
+    private List<Wheel> clonedWheels(List<Wheel> wheels) {
+        List<Wheel> clonedWheels = new ArrayList<>();
```

Comment:
```suggestion
        List<Wheel> clonedWheels = new ArrayList<>(wheels.size());
```
if the default capacity will be more than 10, it is good to pass it to constructor